src/softmax.jl

```diff
@@ -33,20 +33,33 @@ julia> softmax([1, 2, 3])
 See also [`logsoftmax`](@ref).
 """
 softmax(x; dims = 1) = softmax!(similar(x, (float ∘ eltype)(x)), x; dims = dims)
+
 softmax!(x; dims = 1) = softmax!(x, x; dims = dims)
+
 function softmax!(out::O, x::T; dims = 1) where {O<:AbstractArray,T<:AbstractArray}
     out .= exp.(x .- maximum(x; dims = dims))
     out ./= sum(out; dims = dims)
 end
 
-∇softmax(Δ, x; dims = 1) = ∇softmax!(similar(Δ), Δ, x; dims = dims)
-∇softmax!(Δ, x; dims = 1) = ∇softmax!(Δ, Δ, x; dims = dims)
-function ∇softmax!(out::O, Δ::O, x::T; dims = 1) where {O<:AbstractArray,T<:AbstractArray}
-    softmax!(out, x; dims = dims)
-    out .*= Δ .- sum(Δ .* out; dims = dims)
+∇softmax(Δ, x, y; dims = 1) = ∇softmax!(similar(Δ), Δ, x, y; dims = dims)
```

Comment:
```suggestion
∇softmax(Δ:: AbstractArray{T}, x:: AbstractArray, y:: AbstractArray{S}; dims = 1) where {T,S} = 
    ∇softmax!(similar(y, promote_type(T, S)), Δ, x, y; dims = dims)
```
Sometimes you might get `Δ::Fill` or `Δ::BitArray`, maybe not so common but may as well.